iree/compiler/Dialect/Stream/Transforms/AnnotateDispatchArguments.cpp

```diff
@@ -0,0 +1,468 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <memory>
+#include <utility>
+
+#include "iree/compiler/Dialect/Stream/IR/StreamOps.h"
+#include "iree/compiler/Dialect/Stream/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Stream/Transforms/Passes.h"
+#include "iree/compiler/Dialect/Util/Analysis/DFX/Solver.h"
+#include "iree/compiler/Dialect/Util/Analysis/DFX/State.h"
+#include "iree/compiler/Dialect/Util/Analysis/Explorer.h"
+#include "llvm/ADT/APInt.h"
+#include "llvm/Support/Alignment.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/AsmState.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Diagnostics.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/Pass/Pass.h"
+
+#define DEBUG_TYPE "iree-stream-annotate-dispatch-arguments"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Stream {
+namespace {
+
+//===----------------------------------------------------------------------===//
+// Analysis state
+//===----------------------------------------------------------------------===//
+
+// TODO(benvanik): move to Util/Analysis/ as this would be useful in other
+// passes as well and only depends on util.align and upstream ops.
```

Comment:
Seems like we're accumulating a decent chunk of code that _could_ move upstream (Explorer, alignment ops/analysis, etc.).  Could be worth moving the code and setting up namespaces etc. such that adding IREE dependencies to that code is harder. That would also make it easier for people working on other projects to find the code and adapt it to their uses / voice support for contributing it upstream.

Definitely easier to iterate on the code when it exists in the IREE repo directly though :D

Reply:
Stella may be looking into getting align stuff out more - but the explorer is IREE-only (requires tied ops and such) until those interfaces are also upstreamed (if ever) - lots of potential good stuff to take up there, but so little time and given what I see of other people using MLIR I doubt anyone would know what to do with it yet anyway until we can really show it off ;(